timemachine/md/hrex.py

```diff
@@ -0,0 +1,247 @@
+
+    Returns
+    -------
+    NDArray
+        (iter, replica, state) -> cumulative occupancy count: int
+    """
+    replica_idx_by_state_by_iter_ = np.array(replica_idx_by_state_by_iter)  # (iter, state) -> replica
+    _, n_states = replica_idx_by_state_by_iter_.shape
+    states = np.arange(n_states)
+    replica_in_state = replica_idx_by_state_by_iter_[:, :, None] == states  # (iter, replica, state) -> bool
+    cumulative_count = np.cumsum(replica_in_state.astype(int), axis=0)  # (iter, replica, state) -> int
+    return cumulative_count
+
+
+def estimate_transition_matrix(replica_idx_by_state_by_iter: Sequence[Sequence[ReplicaIdx]]) -> NDArray:
+    """Given a mapping of state index to replica index by iteration, returns an estimate of the transition matrix.
+
+    Returns
+    -------
+    NDArray
+        (state, state) -> transition_rate: float
+    """
+    replica_idx_by_state_by_iter_ = np.array(replica_idx_by_state_by_iter)  # (iter, state) -> replica index
+    n_iters, _ = replica_idx_by_state_by_iter_.shape
+
+    # transition_by_iter: (n_iters, n_states, n_states) -> bool
+    transition_by_iter = replica_idx_by_state_by_iter_[:-1, None, :] == replica_idx_by_state_by_iter_[1:, :, None]
+
+    transition_count = np.sum(transition_by_iter, axis=0)  # (to state, from state) -> int
+    transition_rate = transition_count / n_iters  # (to state, from state) -> float
+
+    return transition_rate
+
+
+def estimate_relaxation_time(transition_matrix: NDArray) -> float:
+    """Estimate the relaxation time of permutation moves (in number of iterations) as a function of the second-largest
+    eigenvalue of the transition matrix."""
+
+    eigvals_ascending = np.linalg.eigvals(transition_matrix)
+    mu_2 = eigvals_ascending[-2]  # second-largest eigenvalue
+    return 1 / (1 - mu_2)
+
+
+@dataclass
+class HrexDiagnostics:
+    replica_idx_by_state_by_iter: List[List[ReplicaIdx]]
+    fraction_accepted_by_pair_by_iter: List[List[Tuple[int, int]]]
+
+    @property
+    def cumulative_swap_acceptance_rates(self) -> NDArray:
+        n_accepted, n_proposed = np.moveaxis(np.array(self.fraction_accepted_by_pair_by_iter), -1, 0)
+        return np.cumsum(n_accepted, axis=0) / np.cumsum(n_proposed, axis=0)
+
+    @property
+    def cumulative_replica_state_counts(self) -> NDArray:
+        return get_cumulative_replica_state_counts(self.replica_idx_by_state_by_iter)
+
+    @property
+    def transition_matrix(self) -> NDArray:
+        return estimate_transition_matrix(self.replica_idx_by_state_by_iter)
+
+    @property
+    def relaxation_time(self):
+        return estimate_relaxation_time(self.transition_matrix)
+
+
+def get_swap_attempts_per_iter_heuristic(n_states: int) -> int:
+    """Heuristic for number of swap attempts per iteration derived from [1].
+
+    References
+    ----------
+    [1]: http://dx.doi.org/10.1063/1.3660669
```

Comment:
Should we point at the arvix (https://arxiv.org/abs/1105.5749) rather than the doi? Not sure about this, just nice to be able to get the paper immediately without having to go through the process of getting the paper through the journal.

Reply:
Not sure. I see precedent for both in the repo currently. For published papers it seemed appropriate to list the DOI, but it does make it harder to quickly get a PDF. @maxentile do you have an opinion here?

Reply:
No preference -- I don't think I've used a consistent style.

(When adding references to code I like to include [author, year] title - description of relevance, e.g. https://github.com/proteneer/timemachine/blob/3984fe517d92904986614515810ba392dc60f5a4/timemachine/md/smc.py#L148-L149 )

Reply:
Added arxiv URL additionally in https://github.com/proteneer/timemachine/pull/1128/commits/df187e38c69505b79471a5730fd0037ce509ea2b